### ISSUE TITLE
ofxAudioUnitTap 

### ISSUE THREAD

**Issue by stephanschulz** [roymacdonald/ofxSoundObjects#27]
This add-on looks great.
I hope I can adapt my old code that used the ofxAudioUnit add-on to work with this one.

In ofxAudioUnit I had a ofxAudioUnitTap function available. Is there something similar in this add-on?

I basically have 12 inputs (10 files and 2 mono mics) and output those to 6 mono outputs. 
In my old code I set up a matrix where each file could be patched - via the tap object - to any output.
I see that the "example-soundPlayerMultiOutput" shows how to route inputs to certain outputs. But I have a feeling this is defined during setup and can not be changed on the fly ... ?

Thanks for your advice.

**Comment by roymacdonald** [comment]
Hey Stephan! Thanks.
You should be able to do such with this addon.
You can use whats shown in the  "example-soundPlayerMultiOutput". Any change can be done on the fly. The whole point of this addon is to provide a modular and extendable architecture that can be changed on runtime. 
If you want to change the configuration of the a certain channel group make sure you delete it first from the multiplexer before reconnecting it, otherwise you'll end up with copied stuff.
Take a look at ofxSoundMultiplexer.h for the available functions.

As for what you need to do I would guess that you need to mix the inputs in a certain way before outputting these. So I would go for making some mixer objects where you mix and then connect these to each channel. 
I think that I can make a matrix mixer and add it. Would it be of use to you?

**Comment by stephanschulz** [comment]
Hey Roy.

Yes it would be great to have an addition similar to this class
https://github.com/admsyn/ofxAudioUnit/blob/master/src/ofxAudioUnitMatrixMixer.cpp

I will try your suggestions too.

**Comment by roymacdonald** [comment]
Hi,
I will work on it later and let you know.
do you have an example of how to use ofxAudioUnitMixer? or some code showing how you are using it?
cheers

**Comment by stephanschulz** [comment]
That is awesome.

Here is a screen shot of what my matrix interface looked like. 
on the left you see all the loaded files playing. on the bottom all output channels. the rectangles in the rest of the screen indicate how the routing is done.
<img width="1628" alt="Screen Shot 2018-04-30 at 4 02 52 AM" src="https://user-images.githubusercontent.com/1054816/56808518-b30ff680-67ff-11e9-9c78-a891dbc080e5.png">

here is the full project which might be too much to sift through.
https://gist.github.com/stephanschulz/c6bc8243a1450b90b7e80e0917494118

line 401 in ofApp.mm connects each file to a tap
```
filePlayers[i].connectTo(taps[i+1]);
```

```updateAllMatrixMixerLevels()``` causes the set patch pattern ( levels[][] ) to be applied.

function ```setCrossPoint(int _inputBuss, int _outputChannel, float _level)``` defines which input goes to which output.

please let me know if you need any other info.

thanks a bunch.

**Comment by stephanschulz** [comment]
I guess an other question would be how to feed the 2 channel mic input in to the same mixer matrix.

In example-soundPlayerMultiOutput you define .setInDevice but I don't know how to set it's output channel.

thx

**Comment by roymacdonald** [comment]
Hi
I have not been able to get to work on this today. But I'll get on it soon.

Take a look at the "example-multiInputMixdown" example.
What you need is like a mix between these two examples.

As for your question, the out device is set in
https://github.com/roymacdonald/ofxSoundObjects/blob/master/example-soundPlayerMultiOutput/src/ofApp.cpp#L70

or did you meant a different example?



*Roy J. Macdonald*

**Comment by stephanschulz** [comment]
yes it would super if you have time to add the matrix mixer functionality, where files and mic inputs can be routed around. 
have a great weekend.
